armada_jupyter/submissions.py

```diff
@@ -0,0 +1,205 @@
+
+        ValueError("Timeout must be in the form of [x]h, [x]m or [x]s")
+        return None
+
+    def resources_from_dict(
+        self, resources: Dict[str, Dict[str, Union[str, int]]]
+    ) -> None:
+        """
+        Used for adding a resources from dict
+        """
+
+        self.resources = k8s_resources_from_dict(
+            limits=resources.get(YMLSTR.limits),
+            requests=resources.get(YMLSTR.requests),
+        )
+
+    def __repr__(self) -> str:
+        return (
+            f"Submission(f{YMLSTR.name}='{self.name}', {YMLSTR.image}='{self.image}' "
+            f"{YMLSTR.armada_queue}='{self.armada_queue}', "
+            f"{YMLSTR.armada_priority}={self.armada_priority}, "
+            f"{YMLSTR.timeout}='{self.timeout}', {YMLSTR.resources}={self.resources})"
+        )
+
+
+def k8s_resources_from_dict(
+    limits: Optional[Dict[str, Union[str, int]]] = None,
+    requests: Optional[Dict[str, Union[str, int]]] = None,
+) -> K8sResources:
+    """
+    Creates a K8sResources object from a dict
+    """
+
+    changed: Dict[str, Optional[K8sResourceOptions]] = {}
+
+    if limits is None and requests is None:
+        raise ValueError("Must specify at least one of limits or requests")
+
+    # Use a loop to save repeated code.
+    for key, needs_changed in [(YMLSTR.limits, limits), (YMLSTR.requests, requests)]:
+        if needs_changed is None:
+            changed[key] = None
+            continue
+
+        # We use .get() here because we don't want to raise an error if the key
+        # doesn't exist.
+        changed[key] = K8sResourceOptions(
+            cpu=int(needs_changed.get(YMLSTR.cpu, DEFAULT_CPU)),
+            memory=str(needs_changed.get(YMLSTR.memory, DEFAULT_MEMORY)),
+            nvidia_gpu=needs_changed.get(YMLSTR.nvidia_gpu, None),
+            amd_gpu=needs_changed.get(YMLSTR.amd_gpu, None),
+        )
+
+    return K8sResources(
+        limits=changed[YMLSTR.limits],
+        requests=changed[YMLSTR.requests],
+    )
+
+
+def get_submissions(file) -> List[Submission]:
+    """
+    Creates the submissions from a YAML file
+    """
+
+    submissions = []
+
+    with open(file, "r", encoding="utf-8") as f:
+        config = yaml.load(f, Loader=yaml.FullLoader)
+
+        for submission in config[YMLSTR.submissions]:
+
+            # pop out resources key
+            resources = submission.pop(YMLSTR.resources, None)
+
+            # check for missing mandatory keys
+            for key in [YMLSTR.name, YMLSTR.image]:
+                if key not in submission:
+                    raise ValueError(f"Missing mandatory key {key}")
+
+            # create submission object
+            # We use .get() for the optional keys so that we don't raise an error
+            submission = Submission(
+                name=submission[YMLSTR.name],
+                image=submission[YMLSTR.image],
+                armada_queue=submission.get(YMLSTR.armada_queue, DEFAULT_QUEUE),
```

Comment:
I am not 100% where we should place these defaults, whether we do it in the functions or in the classes. Open to opinions.

Reply:
I am going to leave the defaults here for now. I will write an issue as currently this implementaition will not accept optional args in the classes themselves.